aptos-move/aptos-transaction-simulation/src/state_store.rs

```diff
@@ -0,0 +1,447 @@
+///
+/// This is useful as a base state for situations where no prior state exists.
+#[derive(Debug, Clone)]
+pub struct EmptyStateView;
+
+impl TStateView for EmptyStateView {
+    type Key = StateKey;
+
+    fn get_state_value(&self, _state_key: &Self::Key) -> StateViewResult<Option<StateValue>> {
+        Ok(None)
+    }
+
+    fn get_usage(&self) -> StateViewResult<StateStorageUsage> {
+        Ok(StateStorageUsage::Untracked)
+    }
+
+    fn contains_state(&self, _state_key: &Self::Key) -> StateViewResult<bool> {
+        Ok(false)
+    }
+}
+
+/***************************************************************************************************
+ * Either State View
+ *
+ **************************************************************************************************/
+/// Provides a way to dispatch between two types of state views.
+#[derive(Debug, Clone)]
+pub enum EitherStateView<L, R> {
+    Left(L),
+    Right(R),
+}
+
+impl<L, R, K> TStateView for EitherStateView<L, R>
+where
+    L: TStateView<Key = K>,
+    R: TStateView<Key = K>,
+{
+    type Key = K;
+
+    fn id(&self) -> StateViewId {
+        match self {
+            Self::Left(l) => l.id(),
+            Self::Right(r) => r.id(),
+        }
+    }
+
+    fn get_state_value(&self, state_key: &Self::Key) -> StateViewResult<Option<StateValue>> {
+        match self {
+            Self::Left(l) => l.get_state_value(state_key),
+            Self::Right(r) => r.get_state_value(state_key),
+        }
+    }
+
+    fn get_usage(&self) -> StateViewResult<StateStorageUsage> {
+        match self {
+            Self::Left(l) => l.get_usage(),
+            Self::Right(r) => r.get_usage(),
+        }
+    }
+
+    fn contains_state(&self, state_key: &Self::Key) -> StateViewResult<bool> {
+        match self {
+            Self::Left(l) => l.contains_state(state_key),
+            Self::Right(r) => r.contains_state(state_key),
+        }
+    }
+}
+
+/***************************************************************************************************
+ * Delta State Store
+ *
+ **************************************************************************************************/
+/// A state storage that allows changes to be stacked on top of a base state view.
+///
+/// This is useful for staging reversible state changes or performing simulations on top of
+/// remote states.
+pub struct DeltaStateStore<V> {
+    base: V,
+    states: Mutex<HashMap<StateKey, Option<StateValue>>>,
+}
+
+impl<V> TStateView for DeltaStateStore<V>
+where
+    V: TStateView<Key = StateKey>,
+{
+    type Key = StateKey;
+
+    // TODO: fn id()?
+
+    fn get_state_value(&self, state_key: &Self::Key) -> StateViewResult<Option<StateValue>> {
+        if let Some(res) = self.states.lock().get(state_key) {
+            return Ok(res.clone());
+        }
+        self.base.get_state_value(state_key)
+    }
+
+    fn get_usage(&self) -> StateViewResult<StateStorageUsage> {
+        Ok(StateStorageUsage::Untracked)
```

Comment:
Should it return an error? The base usage exists, but here we added changes on top, not sure if it is better to error or use untracked

Reply:
Not very sure about this one, but changed it to `unimplemented!("not supported")` to start stricter -- easier to relax later.

Reply:
Well, `unimplemented!("not supported")` doesn't work -- a few tests depend on this returning something. I'll keep `Ok(StateStorageUsage::Untracked)` for now, which IMO is reasonable.